[PM-35265] Migration Guide Draft: Push Based Event Delivery (#198)
bitwarden/splunk: docs/migration-guide.md

# Migration Guide: Splunk Application to Event Push Delivery

> This guide will outline the process that the Splunk admin will need to follow to successfully upgrade the Bitwarden Splunk application to receive events from their Bitwarden Organizations using a push based model.

Reviewer: This document should capture our current understanding of what is involved to migrate but currently reads like this document is about the process rather than what it is.

Please review the document for clarity and concision using the following:

- Use present tense
- Conciseness: "will outline the process" -> "documents how to migrate"
- Precision: "upgrade" -> "migrate" Splunk apps can be upgraded/updated but this document covers changes in configuration that may not need a Splunk update (client decides to migrate after a few versions have been released

Example: 
```suggestion
This guide outlines how a Splunk admin migrates the [Bitwarden Event Logs](https://splunkbase.splunk.com/app/6592) Splunk application from polling events to pushed events from their Bitwarden Organizations.
```

Author: Addressed in latest revisions

> Past versions of the application used a polling based model, where the Splunk application used repeated calls to the Bitwarden API for ingesting event data. After application installation, this required the admin to complete a setup form to enable making authenticated HTTP requests to the Bitwarden API. This polling based model will be deprecated in the newest version of the application, in favor of the new push based model.

Reviewer: 🤔 I don't think we decided the polling will be officially deprecated when push driven events are supported. We should update this verbiage or remove it. 
This polling based mode

> This polling based model will be deprecated in the newest version of the application, in favor of the new push based model.

Author: Removed in latest

> Users will be encouraged to complete setup for the new push based model, where the Bitwarden platform will send events into Splunk via the HTTP Event Collector ([see docs](https://help.splunk.com/en/splunk-enterprise/get-started/get-data-in/9.3/get-data-with-http-event-collector/set-up-and-use-http-event-collector-in-splunk-web)).

Reviewer: 👍 Good reference to include

## Upgrading the Application

Splunk admins will be able to find the Bitwarden application through Splunkbase, where the latest version update will be available for download. This download happens in place on their instance and does not require uninstalling their existing version of the Bitwarden Splunk application.

Once the application has been updated, **no changes to existing polling setups for receiving events will occur on their own**. This means that when opening the upgraded application, any existing event polling setups will continue to run as they did on previous versions of the app.

## Complete the Setup for Push Delivery

> While existing event polling setups will still work, the application admin will be prompted to set up push based event delivery. This process includes a new setup form, which will ensure that the Splunk instance has HEC enabled, a token has been generated for use, and the proper setup in the Bitwarden Admin Console has been completed.

Reviewer: 🤔 This reads like we will be pushing the user to set up the push driven events. Rather than prompting, we are exposing the ability to "opt-in".

Author: This has been removed/revised in latest

> After following this setup form and entering the HEC endpoint and token into Bitwarden's Admin Console, the user will be prompted to disable their old polling configuration. Upon confirmation, the application will stop polling for events. Note that duplicate events for the Bitwarden Organization will be received in Splunk for the entire timespan where both push and polling based event delivery models are enabled. **For this reason, it is strongly recommended that the admin disable the polling setup immediately after confirmation that events are being received with the push based model.**

Reviewer: 🤔 Possibly confusing verbiage. Admins cannot specifically disable the polling since the app sets up and manages the polling. Is polling paused when the user navigates to the settings page? Is it paused only when the setup page is complete and settings are saved? How can the user know when push based is on and polling is on? 

The bold note sounds like there is an option to NOT disable the polling setup. This guide should explain that it's an important step in the process and will cause duplicate events if the step is missed and the setup page is not saved.

Author: This was written under the assumption that we would not want to automatically disable existing event polling setups under any conditions. Instead, I was thinking we would expose a way for the app admin to click a button to disable/delete their old polling setup. This way the admin is in control of when polling for events stops, and we don't stop it for them.

Another possibility we could look into would be disabling polling after the admin has successfully set up receiving events via push, but that gets a little trickier since ideally we wouldn't want to disable polling until after we verify that events are coming in properly via push

Author: Alternatively, if we find we prefer allowing gaps in events as opposed to duplicate events, we could make disabling polling a prerequisite to OR part of the set up for configuring push events

Author: Called out as TODO in latest revision. I think we need to make a call as a team on how we want to handle removing old polling configurations

# Migration Guide: Event Polling to Push Delivery

This guide outlines how a Splunk admin can migrate the [Bitwarden Event Logs](https://splunkbase.splunk.com/app/6592) Splunk application from polling events to pushed events from their Bitwarden Organizations.

Past versions of the Splunk application exclusively used a polling model, where the application used repeated calls to the Bitwarden API to retrieve event data. New versions of the application include the ability to retrieve events pushed from the Bitwarden platform into a HTTP Event Collector exposed on the Splunk instance ([see docs](https://help.splunk.com/en/splunk-enterprise/get-started/get-data-in/9.3/get-data-with-http-event-collector/set-up-and-use-http-event-collector-in-splunk-web)).

## Migration Steps

> In order migrate Bitwarden Organization logs from a polling configuration to a push based configuration, admins will complete the following (each step detailed below):

Reviewer: ⚠️ Missing word

```suggestion
In order to migrate Bitwarden Organization logs from a polling configuration to a push based configuration, admins will complete the following (each step detailed below):
```

1. Update the Bitwarden Splunk application

> 2. Complete set up for event push delivery

Reviewer: 🤔  I think it would be worth it to breakdown the Complete setup into two sections to cover the Bitwarden Admin console step and the Bitwarden Event logs separately so it's more clear that the steps are completed in different locations. Though, this would require a change to the sections.

```suggestion
1. Update the Bitwarden Splunk application
2. Bitwarden Admin Console: Complete set up for event push delivery
3. Bitwarden Event Logs Splunk App: Complete setup
```

Suggestion of content for **2. Bitwarden Admin Console: Complete set up for event push delivery**
- A note on step 2 that it's while it's possible to complete step 3 first, we recommend completing step 2 due to the potential loss of events mentioned in https://github.com/bitwarden/splunk/pull/198/changes#r3311550440
- A placeholder link to the Bitwarden Help Center to be filled when a guide is created

Author: @Banrion I swapped the order of the last two steps here. Since the setup flow will start in the Splunk application and be finished in the Admin Console, I think that order makes more sense.

> 2. Complete set up for event push delivery

Reviewer: 🎨 Link to sections below

### Update the Bitwarden Splunk application

Splunk admins need to update to any version after [TODO: add version number here] through Splunkbase. The download happens in place on their instance and does not require uninstalling their existing version of the Bitwarden Splunk application.

> Updating the application will not effect existing event polling configurations, events will continue to be polled just as before the update.

Reviewer: ⚠️ Typo

```suggestion
Updating the application will not affect existing event polling configurations, events will continue to be polled just as before the update.
```

### Complete set up for event push delivery

In the application, a new form for configuring event push delivery is present. This set up will ensure that the Splunk instance has HEC enabled, a token has been generated for use, and proper set up in the Bitwarden Admin Console has been completed.

The admin will complete this form, and the Bitwarden platform will begin to push event logs for the Organization into Splunk.

> **Note that completing the setup form for event push delivery (push or poll) will automatically disable any existing polling configuration.** For example, if your existing configuration uses polling to retrieve events, completing the setup form for event push delivery will overwrite the old polling configuration. Having both event polling and push configurations is not allowed, in order to prevent the same events from being received in Splunk multiple times.

Reviewer: ❓ Ambiguity note. The note reads like checking polling in the setup will disable polling altogether. There won't be a way to complete the setup form for event push delivery AND polling as noted in the example. Specifically the **"(push or poll)"** is confusing. 

I suggest making a new section after Migration Steps to cover this situation:

## Important Notes

### Event Availability in Splunk
Explanation for when events are pushed or polled, covering when events should appear in Splunk based on the different configurations

### Duplicate or Lost Events
Explanation on potential duplicate events, possible lost events if polling is disabled (push driven enabled in the Splunk App first and followed by configuration for push driven events enabled in Admin Console)

Author: Added and filled in recommended sections

Updating the application will not effect existing event polling configurations, events will continue to be polled just as before the update.

### Complete set up for event push delivery

In the application, a new form for configuring event push delivery is present. This set up will ensure that the Splunk instance has HEC enabled, a token has been generated for use, and proper set up in the Bitwarden Admin Console has been completed.

The admin will complete this form, and the Bitwarden platform will begin to push event logs for the Organization into Splunk.

### Disable event polling configurations

Last, it is important to ensure any remaining polling configurations are removed from the application. This prevents the retrieval of duplicate event logs for the Organization, and should be completed as soon as possible. When both poll and push configurations are enabled for an Organization at the same time, the same events will be ingested twice.

> [TODO: clarify on how we would like to build disabling polling configurations. should the admin delete them manually (i.e. button click), or should completing push based delivery delete polling configurations automatically?)]

Reviewer: We can remove this todo. The instructions here should cover the disabling the polling by going to the Bitwarden Event Logs Splunk app setup page and selecting the appropriate setting (see the splunk setup [page documentation](https://docs.splunk.com/Documentation/AddonBuilder/4.5.1/UserGuide/Createasetuppage) on how we are expecting to add the setting via a checkbox)

Author: It sounds to me like we will be automatically disabling polling if the setup form is completed for pushing. I have updated the section to reflect this. Let me know if this isn't quite what you were envisioning.

3. [Bitwarden Admin Console: Complete setup for event push delivery](#bitwarden-admin-console-complete-set-up-for-event-push-delivery)

### Update the Bitwarden Splunk application

Splunk admins need to update to any version after [TODO: add version number here] through Splunkbase. The download happens in place on their instance and does not require uninstalling their existing version of the Bitwarden Splunk application.

Updating the application will not affect existing event polling configurations, events will continue to be polled just as before the update.

### Bitwarden Splunk application: Configure event push delivery (HEC setup)

Once in the updated application, the admin should navigate to the setup form. The setup form includes an option for "push" event delivery. Select it, and the form will assist the admin in ensuring that their Splunk instance is properly configured to receive events using the Http Event Collector (HEC).

Take note of both the HEC endpoint and authentication token, and proceed to the next step where push delivery is configured in the Bitwarden Admin Console.

> ### Bitwarden Admin Console: Complete set up for event push delivery

Reviewer: 🤏 Minor note that this "set up" is inconsistent with "setup" used in other parts of the doc.
